orthanc-gen3/authz_filter.py

```diff
@@ -0,0 +1,94 @@
+"""
+TODO:
+- We are checking access to studies atm, and not access to projects, because we
+don't know which project a study belongs to. This causes duplicate authz
+information in the user.yaml. Kind of hacky option: we could query peregrine -
+if the user can see the study in peregrine, then they have "read" access to
+the project. Or we could query the sheepdog DB directly.
+- When using the Orthanc UI to upload files, this plugin is causing a "failed
+csrf check" error. But we can still upload using the API.
+"""
+
+
+from cachelib import SimpleCache
+from cdislogging import get_logger
+from gen3authz.client.arborist.client import ArboristClient
+import orthanc
+
+
+# Cache the access for 1 second so that we don't make multiple requests to
+# Arborist when a user accesses a webpage and fetches multiple JS/CSS files.
+ACCESS_CACHE = SimpleCache(default_timeout=1)
+
+
+def get_user_jwt(request):
+    """
+    Get the JWT from the request headers. If no JWT is provided, only public
+    data will be available.
+    """
+    user_jwt = request.get("headers", {}).get("access_token")
+    if not user_jwt:
+        user_jwt_parts = request.get("headers", {}).get("authorization", "").split(" ")
+        user_jwt = user_jwt_parts[1] if len(user_jwt_parts) > 1 else None
+    return user_jwt
+
+
+def authorize_user(jwt, service, method, resource):
+    """
+    Return True if the user is authorized to access the resource, False
+    otherwise. If the access is not already cached, ask Arborist.
+    """
+    cache_key = f"{jwt}_{method}_{resource}"
+    if ACCESS_CACHE.has(cache_key):
+        authorized = ACCESS_CACHE.get(cache_key)
+    else:
+        authorized = arborist_client.auth_request(
+            jwt=jwt,
+            service=service,
+            methods=[method],
+            resources=[resource],
+        )
+        ACCESS_CACHE.set(cache_key, authorized)
+    return authorized
+
+
+def check_authorization(uri, **request):
+    """
+    Return True if the user is authorized to access the URI, False otherwise.
+    """
+    service = "dicom-viewer"
+    resource = ""
+    method = ""
+
+    if uri.startswith("/dicom-web/studies/"):
```

Comment:
is it /dicom-web/studies or /dicom-viewer/studies?

Reply:
`/dicom-web/studies` is right, that's the URI Orthanc receives - i think the complete URL is `https://<hostname/dicom-viewer/dicom-web/studies/`